File: app/src/main/java/com/qbapps/claudeusage/domain/guardrail/SessionGuardrailEvaluator.kt
package com.qbapps.claudeusage.domain.guardrail

import com.qbapps.claudeusage.domain.model.UsageHistoryPoint
import com.qbapps.claudeusage.domain.model.UsageMetric
import java.time.Duration
import java.time.Instant
import kotlin.math.abs
import kotlin.math.roundToLong

enum class SessionGuardrailState {
    SAFE,
    STEADY,
    WATCH,
    HIGH,
    CRITICAL,
}

enum class PaceTrack {
    ABOVE_USUAL,
    ON_TRACK,
    BELOW_USUAL,
    UNKNOWN,
}

enum class BurnPhase {
    EARLY,
    MID,
    LATE,
    UNKNOWN,
}

data class SessionGuardrailInsights(
    val state: SessionGuardrailState,
    val currentUtilization: Double?,
    val timeToResetMinutes: Long?,
    val paceTrack: PaceTrack,
    val baselineAtThisPoint: Double?,
    val predictedTimeToCapMinutes: Long?,
    val willHitCapBeforeReset: Boolean,
    val typicalBurnPhase: BurnPhase,
    val earlyHeavyUsageDetected: Boolean,
    val resetReliefSoon: Boolean,
) {
    companion object {
        fun empty(): SessionGuardrailInsights = SessionGuardrailInsights(
            state = SessionGuardrailState.SAFE,
            currentUtilization = null,
            timeToResetMinutes = null,
            paceTrack = PaceTrack.UNKNOWN,
            baselineAtThisPoint = null,
            predictedTimeToCapMinutes = null,
            willHitCapBeforeReset = false,
            typicalBurnPhase = BurnPhase.UNKNOWN,
            earlyHeavyUsageDetected = false,
            resetReliefSoon = false,
        )
    }
}

object SessionGuardrailEvaluator {
    private val sessionDuration: Duration = Duration.ofHours(5)
    private const val sessionMinutes = 300L
    private const val paceBaselineWindowMinutes = 12L
    private const val minBaselineSamples = 6
    private const val minSlopeWindowMinutes = 10L

    fun evaluate(
        currentMetric: UsageMetric?,
        history: List<UsageHistoryPoint>,
        now: Instant = Instant.now(),
    ): SessionGuardrailInsights {
        if (currentMetric == null) return SessionGuardrailInsights.empty()

        val currentUtilization = currentMetric.utilization.coerceIn(0.0, 100.0)
        val resetsAt = currentMetric.resetsAt
        val timeToResetMinutes = resetsAt?.let { Duration.between(now, it).toMinutes().coerceAtLeast(0L) }
        val currentElapsedMinutes = resetsAt?.let { reset ->
            Duration.between(reset.minus(sessionDuration), now).toMinutes().coerceIn(0L, sessionMinutes)
        }

        val allSamples = history.mapNotNull { it.toSessionSample() }
        val currentSessionEpoch = resetsAt?.toEpochMilli()

        val currentSessionSamples = allSamples
            .filter { sample -> sample.sessionResetEpochMs == currentSessionEpoch }
            .plus(currentMetric.toCurrentSample(now))
            .filterNotNull()
            .sortedBy { it.timestamp }

        val previousSessionSamples = allSamples.filter { sample ->
            sample.sessionResetEpochMs != currentSessionEpoch
        }

        val baselineAtThisPoint = currentElapsedMinutes?.let { elapsed ->
            baselineForElapsed(previousSessionSamples, elapsed)
        }
        val paceTrack = when {
            baselineAtThisPoint == null -> PaceTrack.UNKNOWN
            currentUtilization - baselineAtThisPoint >= 8.0 -> PaceTrack.ABOVE_USUAL
            currentUtilization - baselineAtThisPoint <= -8.0 -> PaceTrack.BELOW_USUAL
            else -> PaceTrack.ON_TRACK
        }

        val currentSlope = slopeFromRecentWindow(
            samples = currentSessionSamples,
            referenceElapsedMinutes = currentElapsedMinutes,
        )
        val historicalSlope = averageHistoricalSessionSlope(previousSessionSamples)
        val blendedSlope = when {
            currentSlope != null && historicalSlope != null -> (currentSlope * 0.7) + (historicalSlope * 0.3)
            currentSlope != null -> currentSlope
            else -> historicalSlope
        }

        val predictedTimeToCapMinutes = when {
            currentUtilization >= 100.0 -> 0L
            blendedSlope == null || blendedSlope <= 0.03 -> null
            else -> ((100.0 - currentUtilization) / blendedSlope).roundToLong().coerceAtLeast(0L)
        }

        val willHitCapBeforeReset = predictedTimeToCapMinutes != null &&
            timeToResetMinutes != null &&
            predictedTimeToCapMinutes < timeToResetMinutes

        val typicalBurnPhase = averageBurnByPhase(previousSessionSamples).maxByOrNull { it.value }?.key
            ?: BurnPhase.UNKNOWN

        val earlyHeavyUsageDetected = detectEarlyHeavyUsage(
            currentSessionSamples = currentSessionSamples,
            previousSessionSamples = previousSessionSamples,
            currentElapsedMinutes = currentElapsedMinutes,
            typicalBurnPhase = typicalBurnPhase,
        )

        val resetReliefSoon = currentUtilization >= 75.0 &&
            (timeToResetMinutes != null && timeToResetMinutes <= 45)

        val state = deriveState(
            currentUtilization = currentUtilization,
            timeToResetMinutes = timeToResetMinutes,
            paceTrack = paceTrack,
            predictedTimeToCapMinutes = predictedTimeToCapMinutes,
            willHitCapBeforeReset = willHitCapBeforeReset,
        )

        return SessionGuardrailInsights(
            state = state,
            currentUtilization = currentUtilization,
            timeToResetMinutes = timeToResetMinutes,
            paceTrack = paceTrack,
            baselineAtThisPoint = baselineAtThisPoint,
            predictedTimeToCapMinutes = predictedTimeToCapMinutes,
            willHitCapBeforeReset = willHitCapBeforeReset,
            typicalBurnPhase = typicalBurnPhase,
            earlyHeavyUsageDetected = earlyHeavyUsageDetected,
            resetReliefSoon = resetReliefSoon,
        )
    }

    private fun baselineForElapsed(samples: List<SessionSample>, elapsedMinutes: Long): Double? {
        val matches = samples.filter { sample ->
            abs(sample.elapsedMinutes - elapsedMinutes) <= paceBaselineWindowMinutes
        }
        if (matches.size < minBaselineSamples) return null
        return matches.map { it.utilization }.average()
    }

    private fun slopeFromRecentWindow(
        samples: List<SessionSample>,
        referenceElapsedMinutes: Long?,
    ): Double? {
        if (samples.size < 2 || referenceElapsedMinutes == null) return null

        val recent = samples.filter { sample ->
            referenceElapsedMinutes - sample.elapsedMinutes in 0..90
        }.sortedBy { it.elapsedMinutes }

        if (recent.size < 2) return null
        val first = recent.first()
        val last = recent.last()
        val deltaMinutes = (last.elapsedMinutes - first.elapsedMinutes).coerceAtLeast(0L)
        if (deltaMinutes < minSlopeWindowMinutes) return null
        return (last.utilization - first.utilization) / deltaMinutes.toDouble()
    }

    private fun averageHistoricalSessionSlope(samples: List<SessionSample>): Double? {
        val slopes = samples.groupBy { it.sessionResetEpochMs }
            .values
            .mapNotNull { sessionSamples ->
                val ordered = sessionSamples.sortedBy { it.elapsedMinutes }
                if (ordered.size < 2) return@mapNotNull null
                val first = ordered.first()
                val last = ordered.last()
                val deltaMinutes = (last.elapsedMinutes - first.elapsedMinutes).coerceAtLeast(0L)
                if (deltaMinutes < 20L) return@mapNotNull null
                (last.utilization - first.utilization) / deltaMinutes.toDouble()
            }
        return slopes.takeIf { it.isNotEmpty() }?.average()
    }

    private fun averageBurnByPhase(samples: List<SessionSample>): Map<BurnPhase, Double> {
        val bySession = samples.groupBy { it.sessionResetEpochMs }.values
        if (bySession.isEmpty()) return emptyMap()

        val phaseBurns = mutableMapOf<BurnPhase, MutableList<Double>>(
            BurnPhase.EARLY to mutableListOf(),
            BurnPhase.MID to mutableListOf(),
            BurnPhase.LATE to mutableListOf(),
        )

        bySession.forEach { sessionSamples ->
            val sorted = sessionSamples.sortedBy { it.elapsedMinutes }
            BurnPhase.entries
                .filter { it != BurnPhase.UNKNOWN }
                .forEach { phase ->
                    val burn = burnForPhase(sorted, phase)
                    phaseBurns.getValue(phase).add(burn)
                }
        }

        return phaseBurns.mapValues { (_, values) ->
            values.takeIf { it.isNotEmpty() }?.average() ?: 0.0
        }
    }

    private fun detectEarlyHeavyUsage(
        currentSessionSamples: List<SessionSample>,
        previousSessionSamples: List<SessionSample>,
        currentElapsedMinutes: Long?,
        typicalBurnPhase: BurnPhase,
    ): Boolean {
        if (currentElapsedMinutes == null || currentElapsedMinutes > 120L) return false
        if (typicalBurnPhase == BurnPhase.EARLY || typicalBurnPhase == BurnPhase.UNKNOWN) return false

        val currentEarlyBurn = burnForPhase(currentSessionSamples, BurnPhase.EARLY)
        if (currentEarlyBurn <= 0.0) return false

        val historicalEarlyBurns = previousSessionSamples
            .groupBy { it.sessionResetEpochMs }
            .values
            .map { session -> burnForPhase(session.sortedBy { it.elapsedMinutes }, BurnPhase.EARLY) }
            .filter { burn -> burn > 0.0 }

        val historicalAverage = historicalEarlyBurns.takeIf { it.isNotEmpty() }?.average() ?: return false
        return currentEarlyBurn > historicalAverage + 10.0
    }

    private fun burnForPhase(
        sessionSamples: List<SessionSample>,
        phase: BurnPhase,
    ): Double {
        val range = when (phase) {
            BurnPhase.EARLY -> 0L..60L
            BurnPhase.MID -> 61L..240L
            BurnPhase.LATE -> 241L..300L
            BurnPhase.UNKNOWN -> return 0.0
        }
        val phasePoints = sessionSamples.filter { it.elapsedMinutes in range }
        if (phasePoints.size < 2) return 0.0
        val min = phasePoints.minOf { it.utilization }
        val max = phasePoints.maxOf { it.utilization }
        return (max - min).coerceAtLeast(0.0)
    }

    private fun deriveState(
        currentUtilization: Double,
        timeToResetMinutes: Long?,
        paceTrack: PaceTrack,
        predictedTimeToCapMinutes: Long?,
        willHitCapBeforeReset: Boolean,
    ): SessionGuardrailState {
        val fromUtil = when {
            currentUtilization >= 90.0 -> SessionGuardrailState.CRITICAL
            currentUtilization >= 80.0 -> SessionGuardrailState.HIGH
            currentUtilization >= 65.0 -> SessionGuardrailState.WATCH
            currentUtilization >= 45.0 -> SessionGuardrailState.STEADY
            else -> SessionGuardrailState.SAFE
        }

        var state = fromUtil

        if (paceTrack == PaceTrack.ABOVE_USUAL && state.ordinal < SessionGuardrailState.WATCH.ordinal) {
            state = SessionGuardrailState.WATCH
        }

        if (predictedTimeToCapMinutes != null && timeToResetMinutes != null) {
            when {
                willHitCapBeforeReset -> state = maxState(state, SessionGuardrailState.CRITICAL)
                predictedTimeToCapMinutes <= timeToResetMinutes + 30L -> {
                    state = maxState(state, SessionGuardrailState.WATCH)
                }
            }
        }

        if (!willHitCapBeforeReset &&
            timeToResetMinutes != null &&
            timeToResetMinutes <= 20L &&
            state.ordinal > SessionGuardrailState.HIGH.ordinal
        ) {
            state = SessionGuardrailState.HIGH
        }

        return state
    }

    private fun maxState(a: SessionGuardrailState, b: SessionGuardrailState): SessionGuardrailState {
        return if (a.ordinal >= b.ordinal) a else b
    }

    private fun UsageHistoryPoint.toSessionSample(): SessionSample? {
        val utilizationValue = fiveHourUtilization ?: return null
        val resetAt = fiveHourResetsAt ?: return null
        val elapsedMinutes = Duration.between(resetAt.minus(sessionDuration), timestamp)
            .toMinutes()
            .coerceIn(0L, sessionMinutes)
        return SessionSample(
            timestamp = timestamp,
            utilization = utilizationValue.coerceIn(0.0, 100.0),
            elapsedMinutes = elapsedMinutes,
            sessionResetEpochMs = resetAt.toEpochMilli(),
        )
    }

    private fun UsageMetric.toCurrentSample(now: Instant): SessionSample? {
        val resetAt = resetsAt ?: return null
        val elapsedMinutes = Duration.between(resetAt.minus(sessionDuration), now)
            .toMinutes()
            .coerceIn(0L, sessionMinutes)
        return SessionSample(
            timestamp = now,
            utilization = utilization.coerceIn(0.0, 100.0),
            elapsedMinutes = elapsedMinutes,
            sessionResetEpochMs = resetAt.toEpochMilli(),
        )
    }

    private data class SessionSample(
        val timestamp: Instant,
        val utilization: Double,
        val elapsedMinutes: Long,
        val sessionResetEpochMs: Long,
    )
}

File: ios/OpenUsageCore/Sources/OpenUsageCore/Guardrail/SessionGuardrailEvaluator.swift
import Foundation

/// Faithful port of `domain/guardrail/SessionGuardrailEvaluator.kt`.
/// Ordinal-sensitive comparisons in the Kotlin original are preserved by giving
/// the states explicit `rank` values.

public enum SessionGuardrailState: String, Sendable, CaseIterable {
    case safe
    case steady
    case watch
    case high
    case critical

    /// Equivalent to Kotlin's `enum.ordinal`.
    public var rank: Int {
        switch self {
        case .safe: return 0
        case .steady: return 1
        case .watch: return 2
        case .high: return 3
        case .critical: return 4
        }
    }
}

public enum PaceTrack: String, Sendable {
    case aboveUsual
    case onTrack
    case belowUsual
    case unknown
}

public enum BurnPhase: String, Sendable {
    case early
    case mid
    case late
    case unknown
}

public struct SessionGuardrailInsights: Equatable, Sendable {
    public let state: SessionGuardrailState
    public let currentUtilization: Double?
    public let timeToResetMinutes: Int?
    public let paceTrack: PaceTrack
    public let baselineAtThisPoint: Double?
    public let predictedTimeToCapMinutes: Int?
    public let willHitCapBeforeReset: Bool
    public let typicalBurnPhase: BurnPhase
    public let earlyHeavyUsageDetected: Bool
    public let resetReliefSoon: Bool

    public init(
        state: SessionGuardrailState,
        currentUtilization: Double?,
        timeToResetMinutes: Int?,
        paceTrack: PaceTrack,
        baselineAtThisPoint: Double?,
        predictedTimeToCapMinutes: Int?,
        willHitCapBeforeReset: Bool,
        typicalBurnPhase: BurnPhase,
        earlyHeavyUsageDetected: Bool,
        resetReliefSoon: Bool
    ) {
        self.state = state
        self.currentUtilization = currentUtilization
        self.timeToResetMinutes = timeToResetMinutes
        self.paceTrack = paceTrack
        self.baselineAtThisPoint = baselineAtThisPoint
        self.predictedTimeToCapMinutes = predictedTimeToCapMinutes
        self.willHitCapBeforeReset = willHitCapBeforeReset
        self.typicalBurnPhase = typicalBurnPhase
        self.earlyHeavyUsageDetected = earlyHeavyUsageDetected
        self.resetReliefSoon = resetReliefSoon
    }

    public static func empty() -> SessionGuardrailInsights {
        SessionGuardrailInsights(
            state: .safe,
            currentUtilization: nil,
            timeToResetMinutes: nil,
            paceTrack: .unknown,
            baselineAtThisPoint: nil,
            predictedTimeToCapMinutes: nil,
            willHitCapBeforeReset: false,
            typicalBurnPhase: .unknown,
            earlyHeavyUsageDetected: false,
            resetReliefSoon: false
        )
    }
}

public enum SessionGuardrailEvaluator {
    static let sessionDurationSeconds: TimeInterval = 5 * 60 * 60
    static let sessionMinutes = 300
    private static let paceBaselineWindowMinutes = 12
    private static let minBaselineSamples = 6
    private static let minSlopeWindowMinutes = 10

    struct SessionSample {
        let timestamp: Date
        let utilization: Double
        let elapsedMinutes: Int
        /// Kotlin keys sessions on `resetsAt.toEpochMilli()`.
        let sessionResetEpochMs: Int64
    }

    public static func evaluate(
        currentMetric: UsageMetric?,
        history: [UsageHistoryPoint],
        now: Date = Date()
    ) -> SessionGuardrailInsights {
        guard let currentMetric else { return .empty() }

        let currentUtilization = clamp(currentMetric.utilization, 0.0, 100.0)
        let resetsAt = currentMetric.resetsAt
        let timeToResetMinutes = resetsAt.map { max(minutesBetween(now, $0), 0) }
        let currentElapsedMinutes: Int? = resetsAt.map { reset in
            let start = reset.addingTimeInterval(-sessionDurationSeconds)
            return clampInt(minutesBetween(start, now), 0, sessionMinutes)
        }

        let allSamples = history.compactMap(sessionSample(from:))
        let currentSessionEpoch = resetsAt.map(epochMillis)

        var currentSessionSamples = allSamples.filter { $0.sessionResetEpochMs == currentSessionEpoch }
        if let sample = currentSample(from: currentMetric, now: now) {
            currentSessionSamples.append(sample)
        }
        currentSessionSamples.sort { $0.timestamp < $1.timestamp }

        let previousSessionSamples = allSamples.filter { $0.sessionResetEpochMs != currentSessionEpoch }

        let baselineAtThisPoint = currentElapsedMinutes.flatMap {
            baselineForElapsed(previousSessionSamples, elapsedMinutes: $0)
        }

        let paceTrack: PaceTrack
        if let baselineAtThisPoint {
            if currentUtilization - baselineAtThisPoint >= 8.0 {
                paceTrack = .aboveUsual
            } else if currentUtilization - baselineAtThisPoint <= -8.0 {
                paceTrack = .belowUsual
            } else {
                paceTrack = .onTrack
            }
        } else {
            paceTrack = .unknown
        }

        let currentSlope = slopeFromRecentWindow(
            samples: currentSessionSamples,
            referenceElapsedMinutes: currentElapsedMinutes
        )
        let historicalSlope = averageHistoricalSessionSlope(previousSessionSamples)
        let blendedSlope: Double?
        if let currentSlope, let historicalSlope {
            blendedSlope = (currentSlope * 0.7) + (historicalSlope * 0.3)
        } else if let currentSlope {
            blendedSlope = currentSlope
        } else {
            blendedSlope = historicalSlope
        }

        let predictedTimeToCapMinutes: Int?
        if currentUtilization >= 100.0 {
            predictedTimeToCapMinutes = 0
        } else if let blendedSlope, blendedSlope > 0.03 {
            predictedTimeToCapMinutes = max(Int(((100.0 - currentUtilization) / blendedSlope).rounded()), 0)
        } else {
            predictedTimeToCapMinutes = nil
        }

        var willHitCapBeforeReset = false
        if let predictedTimeToCapMinutes, let timeToResetMinutes {
            willHitCapBeforeReset = predictedTimeToCapMinutes < timeToResetMinutes
        }

        let typicalBurnPhase = dominantBurnPhase(previousSessionSamples)

        let earlyHeavyUsageDetected = detectEarlyHeavyUsage(
            currentSessionSamples: currentSessionSamples,
            previousSessionSamples: previousSessionSamples,
            currentElapsedMinutes: currentElapsedMinutes,
            typicalBurnPhase: typicalBurnPhase
        )

        let resetReliefSoon = currentUtilization >= 75.0 &&
            (timeToResetMinutes != nil && timeToResetMinutes! <= 45)

        let state = deriveState(
            currentUtilization: currentUtilization,
            timeToResetMinutes: timeToResetMinutes,
            paceTrack: paceTrack,
            predictedTimeToCapMinutes: predictedTimeToCapMinutes,
            willHitCapBeforeReset: willHitCapBeforeReset
        )

        return SessionGuardrailInsights(
            state: state,
            currentUtilization: currentUtilization,
            timeToResetMinutes: timeToResetMinutes,
            paceTrack: paceTrack,
            baselineAtThisPoint: baselineAtThisPoint,
            predictedTimeToCapMinutes: predictedTimeToCapMinutes,
            willHitCapBeforeReset: willHitCapBeforeReset,
            typicalBurnPhase: typicalBurnPhase,
            earlyHeavyUsageDetected: earlyHeavyUsageDetected,
            resetReliefSoon: resetReliefSoon
        )
    }

    // MARK: - Private helpers

    private static func baselineForElapsed(_ samples: [SessionSample], elapsedMinutes: Int) -> Double? {
        let matches = samples.filter { abs($0.elapsedMinutes - elapsedMinutes) <= paceBaselineWindowMinutes }
        guard matches.count >= minBaselineSamples else { return nil }
        return matches.map(\.utilization).reduce(0, +) / Double(matches.count)
    }

    private static func slopeFromRecentWindow(
        samples: [SessionSample],
        referenceElapsedMinutes: Int?
    ) -> Double? {
        guard samples.count >= 2, let referenceElapsedMinutes else { return nil }

        let recent = samples
            .filter { (0...90).contains(referenceElapsedMinutes - $0.elapsedMinutes) }
            .sorted { $0.elapsedMinutes < $1.elapsedMinutes }

        guard recent.count >= 2, let first = recent.first, let last = recent.last else { return nil }
        let deltaMinutes = max(last.elapsedMinutes - first.elapsedMinutes, 0)
        guard deltaMinutes >= minSlopeWindowMinutes else { return nil }
        return (last.utilization - first.utilization) / Double(deltaMinutes)
    }

    private static func averageHistoricalSessionSlope(_ samples: [SessionSample]) -> Double? {
        let slopes = groupedBySession(samples).compactMap { sessionSamples -> Double? in
            let ordered = sessionSamples.sorted { $0.elapsedMinutes < $1.elapsedMinutes }
            guard ordered.count >= 2, let first = ordered.first, let last = ordered.last else { return nil }
            let deltaMinutes = max(last.elapsedMinutes - first.elapsedMinutes, 0)
            guard deltaMinutes >= 20 else { return nil }
            return (last.utilization - first.utilization) / Double(deltaMinutes)
        }
        guard !slopes.isEmpty else { return nil }
        return slopes.reduce(0, +) / Double(slopes.count)
    }

    /// Kotlin builds a `LinkedHashMap` keyed EARLY, MID, LATE and takes
    /// `maxByOrNull`, which returns the *first* maximum. An empty sample set
    /// yields an empty map and therefore `UNKNOWN`.
    private static func dominantBurnPhase(_ samples: [SessionSample]) -> BurnPhase {
        let sessions = groupedBySession(samples)
        guard !sessions.isEmpty else { return .unknown }

        let phases: [BurnPhase] = [.early, .mid, .late]
        var averages: [(BurnPhase, Double)] = []
        for phase in phases {
            let burns = sessions.map { session in
                burnForPhase(session.sorted { $0.elapsedMinutes < $1.elapsedMinutes }, phase: phase)
            }
            let average = burns.isEmpty ? 0.0 : burns.reduce(0, +) / Double(burns.count)
            averages.append((phase, average))
        }

        var best = averages[0]
        for candidate in averages.dropFirst() where candidate.1 > best.1 {
            best = candidate
        }
        return best.0
    }

    private static func detectEarlyHeavyUsage(
        currentSessionSamples: [SessionSample],
        previousSessionSamples: [SessionSample],
        currentElapsedMinutes: Int?,
        typicalBurnPhase: BurnPhase
    ) -> Bool {
        guard let currentElapsedMinutes, currentElapsedMinutes <= 120 else { return false }
        if typicalBurnPhase == .early || typicalBurnPhase == .unknown { return false }

        let currentEarlyBurn = burnForPhase(currentSessionSamples, phase: .early)
        guard currentEarlyBurn > 0.0 else { return false }

        let historicalEarlyBurns = groupedBySession(previousSessionSamples)
            .map { burnForPhase($0.sorted { $0.elapsedMinutes < $1.elapsedMinutes }, phase: .early) }
            .filter { $0 > 0.0 }

        guard !historicalEarlyBurns.isEmpty else { return false }
        let historicalAverage = historicalEarlyBurns.reduce(0, +) / Double(historicalEarlyBurns.count)
        return currentEarlyBurn > historicalAverage + 10.0
    }

    private static func burnForPhase(_ sessionSamples: [SessionSample], phase: BurnPhase) -> Double {
        let range: ClosedRange<Int>
        switch phase {
        case .early: range = 0...60
        case .mid: range = 61...240
        case .late: range = 241...300
        case .unknown: return 0.0
        }
        let phasePoints = sessionSamples.filter { range.contains($0.elapsedMinutes) }
        guard phasePoints.count >= 2 else { return 0.0 }
        let minimum = phasePoints.map(\.utilization).min() ?? 0.0
        let maximum = phasePoints.map(\.utilization).max() ?? 0.0
        return max(maximum - minimum, 0.0)
    }

    private static func deriveState(
        currentUtilization: Double,
        timeToResetMinutes: Int?,
        paceTrack: PaceTrack,
        predictedTimeToCapMinutes: Int?,
        willHitCapBeforeReset: Bool
    ) -> SessionGuardrailState {
        var state: SessionGuardrailState
        if currentUtilization >= 90.0 {
            state = .critical
        } else if currentUtilization >= 80.0 {
            state = .high
        } else if currentUtilization >= 65.0 {
            state = .watch
        } else if currentUtilization >= 45.0 {
            state = .steady
        } else {
            state = .safe
        }

        if paceTrack == .aboveUsual, state.rank < SessionGuardrailState.watch.rank {
            state = .watch
        }

        if let predictedTimeToCapMinutes, let timeToResetMinutes {
            if willHitCapBeforeReset {
                state = maxState(state, .critical)
            } else if predictedTimeToCapMinutes <= timeToResetMinutes + 30 {
                state = maxState(state, .watch)
            }
        }

        if !willHitCapBeforeReset,
           let timeToResetMinutes,
           timeToResetMinutes <= 20,
           state.rank > SessionGuardrailState.high.rank {
            state = .high
        }

        return state
    }

    private static func maxState(_ a: SessionGuardrailState, _ b: SessionGuardrailState) -> SessionGuardrailState {
        a.rank >= b.rank ? a : b
    }

    private static func groupedBySession(_ samples: [SessionSample]) -> [[SessionSample]] {
        var order: [Int64] = []
        var buckets: [Int64: [SessionSample]] = [:]
        for sample in samples {
            if buckets[sample.sessionResetEpochMs] == nil {
                order.append(sample.sessionResetEpochMs)
                buckets[sample.sessionResetEpochMs] = []
            }
            buckets[sample.sessionResetEpochMs]?.append(sample)
        }
        return order.compactMap { buckets[$0] }
    }

    static func sessionSample(from point: UsageHistoryPoint) -> SessionSample? {
        guard let utilizationValue = point.fiveHourUtilization,
              let resetAt = point.fiveHourResetsAt else { return nil }
        let start = resetAt.addingTimeInterval(-sessionDurationSeconds)
        let elapsedMinutes = clampInt(minutesBetween(start, point.timestamp), 0, sessionMinutes)
        return SessionSample(
            timestamp: point.timestamp,
            utilization: clamp(utilizationValue, 0.0, 100.0),
            elapsedMinutes: elapsedMinutes,
            sessionResetEpochMs: epochMillis(resetAt)
        )
    }

    private static func currentSample(from metric: UsageMetric, now: Date) -> SessionSample? {
        guard let resetAt = metric.resetsAt else { return nil }
        let start = resetAt.addingTimeInterval(-sessionDurationSeconds)
        let elapsedMinutes = clampInt(minutesBetween(start, now), 0, sessionMinutes)
        return SessionSample(
            timestamp: now,
            utilization: clamp(metric.utilization, 0.0, 100.0),
            elapsedMinutes: elapsedMinutes,
            sessionResetEpochMs: epochMillis(resetAt)
        )
    }
}

// MARK: - Shared numeric helpers

/// `java.time.Duration.between(a, b).toMinutes()` truncates toward zero.
func minutesBetween(_ from: Date, _ to: Date) -> Int {
    Int(to.timeIntervalSince(from) / 60.0)
}

func epochMillis(_ date: Date) -> Int64 {
    Int64((date.timeIntervalSince1970 * 1000.0).rounded())
}

func clamp(_ value: Double, _ low: Double, _ high: Double) -> Double {
    min(max(value, low), high)
}

func clampInt(_ value: Int, _ low: Int, _ high: Int) -> Int {
    min(max(value, low), high)
}
